### ISSUE TITLE
Can we support superagent

### ISSUE THREAD

**Issue by junchaozhang** [elastic/apm-agent-nodejs#897]
We use Node.js to do the API layer, so the important thing for us is to know the performance of the network request. And we use Superagent to send the request.But I can not get any trace in the tracing graph.
So I want to ask: do the elastic apm agent support superagent, if not support, how can I do this by myself

**Comment by watson** [comment]
We normally catch all those 3rd party HTTP clients as we integrate with the core HTTP and HTTPS library in Node.js core. Do you know if Superagent happen to use their own HTTP integration?

**Comment by watson** [comment]
Come to think of it, this could also be a context propagation issue, where Superagent employs its own callback queue that we would need to patch 🤔

If possible, could you try to set the agent `logLevel` to `trace`, send a request to one of the routes where Superagent is supposed to perform an outgoing request and then share the resulting logs here? Then we can see what exactly is going on.

**Comment by junchaozhang** [comment]
thx for your reply, and here is the trace log:
```
intercepted request event call to http.Server.prototype.emit
start trace { trans: '20bdf2e3e8402d23',
  parent: undefined,
  trace: '7d8f05cca18f4214843eed644e434f49',
  name: null,
  type: null }
setting default transaction name: GET /ss { trans: '20bdf2e3e8402d23',
  parent: undefined,
  trace: '7d8f05cca18f4214843eed644e434f49' }
no active transaction found - cannot build new span
intercepted call to http.request { id: null }
setting transaction result { trans: '20bdf2e3e8402d23',
  parent: undefined,
  trace: '7d8f05cca18f4214843eed644e434f49',
  result: 'HTTP 2xx' }
sending transaction { trans: '20bdf2e3e8402d23',
  trace: '7d8f05cca18f4214843eed644e434f49' }
ended transaction { trans: '20bdf2e3e8402d23',
  parent: undefined,
  trace: '7d8f05cca18f4214843eed644e434f49',
  type: 'request',
  result: 'HTTP 2xx',
  name: 'GET /ss' }
```

And I also check the new relic implement:
https://www.npmjs.com/package/@newrelic/superagent?activeTab=dependencies

They bind these 3 method:
```javascript
  if (proto) {
    shim.wrapReturn(proto, 'request', wrapHttpReq)
    shim.wrap(proto, 'callback', wrapCallback)
    shim.wrap(proto, 'then', wrapThen)
  }
```

So I think we also need to bind these 3 methods and write an instrument in modules

**Comment by junchaozhang** [comment]
here is the demo for testing supporting superagent:
```
require('elastic-apm-node').start({
  // Override service name from package.json
  // Allowed characters: a-z, A-Z, 0-9, -, _, and space
  serviceName: 'koa-local',

  // Use if APM Server requires a token
  secretToken: '',

  // Set custom APM Server URL (default: http://localhost:8200)
  serverUrl: 'http://localhost:8200',
  logLevel: 'trace',
});

const Koa = require('koa');
const Router = require('koa-router');
const request = require('superagent');
const https = require('https');

const app = new Koa();
const router = new Router();

router.get('/ss', async (ctx, next) => {
  const response = await request.get('https://petstore.swagger.io/v2/store/inventory');
  ctx.body = response.body;
})

app.use(router.routes())
  .use(router.allowedMethods());

app.listen(3000);

```

And here is the result in apm:
<img width="1246" alt="screen shot 2019-03-05 at 11 37 19 pm" src="https://user-images.githubusercontent.com/6490080/53817175-dce73280-3f9f-11e9-93eb-ec98d906fb10.png">

**Comment by watson** [comment]
Thanks for sharing the logs. Yes, it does look like a context propagation issue as Superagent maintains its own user-land callback queue. That's why you see this in the logs:

```
no active transaction found - cannot build new span
intercepted call to http.request { id: null }
```

Preferably we should open a PR with them to fix that, but we can patch this on our side as well. We can't use the New Relic code as that's not open source. But it shouldn't be that hard I hope - especially with the test example that you shared, thanks 😃

**Comment by junchaozhang** [comment]
> Thanks for sharing the logs. Yes, it does look like a context propagation issue as Superagent maintains its own user-land callback queue. That's why you see this in the logs:
> 
> ```
> no active transaction found - cannot build new span
> intercepted call to http.request { id: null }
> ```
> Preferably we should open a PR with them to fix that, but we can patch this on our side as well. We can't use the New Relic code as that's not open source. But it shouldn't be that hard I hope - especially with the test example that you shared, thanks 😃

Yes, I'm also try to patch it, and I also find that this is the context propagation problem.
So I think we can bind the superagent own user-land callback queue to the current transaction

**Comment by junchaozhang** [comment]
> Thanks for sharing the logs. Yes, it does look like a context propagation issue as Superagent maintains its own user-land callback queue. That's why you see this in the logs:
> 
> ```
> no active transaction found - cannot build new span
> intercepted call to http.request { id: null }
> ```
> Preferably we should open a PR with them to fix that, but we can patch this on our side as well. We can't use the New Relic code as that's not open source. But it shouldn't be that hard I hope - especially with the test example that you shared, thanks 😃

I try to make an instrumentation, and I try it, but it will override the top transaction:
![image](https://user-images.githubusercontent.com/6490080/54123987-1c000280-443c-11e9-8e64-a33681c7afc5.png)

but if I create the span, it can not get the active transaction.

I want to make the superagent be a span of the top transaction,
So I want to ask, can I make the superagent request to be a transaction,
I read the article here:https://github.com/elastic/apm-agent-nodejs/pull/905#issuecomment-470093427
1. Can I make the relation between the top transaction and the superagent trasaction.
2. If I create the span, it can not get the active transaction, how can I get the active transaction

here is my code:
```javascript
'use strict'

var url = require('url')

var eos = require('end-of-stream')

var shimmer = require('../shimmer')
var symbols = require('../../symbols')

module.exports = function (superagent, agent, version, enabled) {
  if (!enabled) return superagent
  var ins = agent._instrumentation
  agent.logger.debug('shimming superagent.createServer function')

  const proto = superagent.Request && superagent.Request.prototype

  shimmer.wrap(proto, 'request', wrapHttpReq)
  shimmer.wrap(proto, 'then', wrapThen)

  return superagent

  function wrapHttpReq(original) {
    return function wrappedHttpReq() {
      var span = agent.startSpan(null,'ext.superagent')
      agent.logger.debug('http request')
      const req = original.apply(this, arguments)

      if(!span) return req
      agent._instrumentation.bindEmitter(req)
      return req
    }
  }

  function wrapThen(original) {
    return function wrappedThen() {
      agent.logger.debug('then')
      
      return new Promise((resolve, reject) => {
        return original.apply(this, arguments).then((result) => {
          var trans = agent._instrumentation.currentTransaction
           resolve(result)
        }, (err) => {
          agent.captureError(err)
          endTransaction(transaction)
          reject(err)
        })
      })
    }
  }
}

```

@watson

**Comment by watson** [comment]
> 1. Can I make the relation between the top transaction and the superagent trasaction.

No, there can be only one transaction active within a service, the outgoing request via superagent has to be implemented as a span.
 
> 2. If I create the span, it can not get the active transaction, how can I get the active transaction

This is most likely due to a context propagation issue where superagent have implemented its own callback queue. Maybe to manage a tcp socket pool.

> 3. In this case, can this be the span of the top transaction, and how to do it

Good question. If this is a context propagation issue caused by superagent, we have to find the callback where the context is lost and use `agent._instrumentation.bindFunction(callback)`  to ensure the context (i.e. the current transaction) is restored correctly once the `callback` is called.

Finding this place can be very tricky. And it might not be in superagent it self, but could easily be in a sub-dependency of superagent.

I'll just take a look quickly to see if I can spot it.

**Comment by junchaozhang** [comment]
> > 1. Can I make the relation between the top transaction and the superagent trasaction.
> 
> No, there can be only one transaction active within a service, the outgoing request via superagent has to be implemented as a span.
> 
> > 1. If I create the span, it can not get the active transaction, how can I get the active transaction
> 
> This is most likely due to a context propagation issue where superagent have implemented its own callback queue. Maybe to manage a tcp socket pool.
> 
> > 1. In this case, can this be the span of the top transaction, and how to do it
> 
> Good question. If this is a context propagation issue caused by superagent, we have to find the callback where the context is lost and use `agent._instrumentation.bindFunction(callback)` to ensure the context (i.e. the current transaction) is restored correctly once the `callback` is called.
> 
> Finding this place can be very tricky. And it might not be in superagent it self, but could easily be in a sub-dependency of superagent.
> 
> I'll just take a look quickly to see if I can spot it.

thanks for your reply, I'm also finding where the context lose. I think is here:
https://github.com/visionmedia/superagent/blob/master/lib/request-base.js#L238

So we can get the resolve and bind it I think

**Comment by junchaozhang** [comment]
and I also find that, the newrelic package wrap these functions:
```javascript
    shim.wrapReturn(proto, 'request', wrapHttpReq)
    shim.wrap(proto, 'callback', wrapCallback)
    shim.wrap(proto, 'then', wrapThen)
```

So I think these function cause the context lose

**Comment by watson** [comment]
I just performed a test with superagent without using Koa, but just using superagent by it self, and we didn't loose context then. So it seems to only be in combination with koa or koa-router. So maybe this is not an issue with superagent at all.

Here's the code I ran:

```js
'use strict'

const agent = require('elastic-apm-node').start({ logLevel: 'trace' })
const request = require('superagent')

asyncAwaitStyle().then(callbackStyle)

async function asyncAwaitStyle () {
  agent.startTransaction()
  console.log('SENDING REQUEST!')

  await request.get('https://petstore.swagger.io/v2/store/inventory')

  console.log('DONE!')
  agent.endTransaction()
}

function callbackStyle () {
  agent.startTransaction()

  console.log('SENDING REQUEST!')
  request.get('https://petstore.swagger.io/v2/store/inventory', function () {
    console.log('DONE!')
    agent.endTransaction()
  })
}
```

**Comment by watson** [comment]
Update: When using callback style for the superagent request instead of async/await in the koa route, it works just fine. Not sure exactly what that means yet, but just wanted to note it down in case it's important

**Comment by watson** [comment]
Update 2: If you set the Elastic APM config option `asyncHooks: false`, it works fine. This might mean it's something with how promises are traced in async hooks

**Comment by junchaozhang** [comment]
> Update 2: If you set the Elastic APM config option `asyncHooks: false`, it works fine. This might mean it's something with how promises are traced in async hooks

1. I try your demo, and it works, I can see there are 2 transactions in the elastic apm.
**I think you are right, it's not the superagent problem, maybe is the combination problem of superagent and koa.**


Here is the log of:
[superagent-async-test.log](https://github.com/elastic/apm-agent-nodejs/files/2957470/superagent-async-test.log)

and I also see has the lines:
```
no active transaction found - cannot build new span
intercepted call to http.request { id: null }
```

So what this means? why this still has no active transaction?

2. I change the  option `asyncHooks: false`, it works, but I want to ask if I change the options, what the good or bad thing will happen 
[Uploading superagent-async-test.log…]()

**Comment by junchaozhang** [comment]
> Update 2: If you set the Elastic APM config option `asyncHooks: false`, it works fine. This might mean it's something with how promises are traced in async hooks

1. I try your demo, and it works, I can see there are 2 transactions in the elastic apm.
**I think you are right, it's not the superagent problem, maybe is the combination problem of superagent and koa.**


Here is the log of:
[superagent-async-test.log](https://github.com/elastic/apm-agent-nodejs/files/2957476/superagent-async-test.log)


and I also see has the lines:
```
no active transaction found - cannot build new span
intercepted call to http.request { id: null }
```

So what this means? why this still has no active transaction?

2. I change the  option `asyncHooks: false`, it works, but I want to ask if I change the options, what the good or bad thing will happen

**Comment by watson** [comment]
> So what this means? why this still has no active transaction?

I think that's actually just the agent seeing it's own request to the APM Server. We could clean that up so it's not as distracting.

> 2. I change the option `asyncHooks: false`, it works, but I want to ask what's the different if I change the option, will I lose a lot performance?

[Async Hooks](https://nodejs.org/api/async_hooks.html) is an API added in Node.js 8 that allows us to know every time an async callback is being queued and then later when it's being called. This allows to store a reference to the transaction that was active when the callback was queued, and then restore it when ever the callback is called.

If you disable our use of that API by using `asyncHooks: false`, we will fall back to patching all async functions in Node core to basically achieve the same thing. E.g: when ever you call `http.request`, you actually call a tiny "middleware" function provided by the APM agent that records the active transaction before proxying the call on to the actual `http.request` function.

So the Async Hooks API is a lot cleaner and should normally have a slightly lower overhead. That being said, there's still some performance issues in Node core regarding Async Hooks and heavy use of Promises. So if your application uses a lot of Promises, you'll most likely see a slight performance boost by setting `asyncHooks: false`. But this is just until the Performance issue is fixed in Node core.

**Comment by watson** [comment]
This issue might be the same as discussed in #873

**Comment by dpjayasekara** [comment]
I am experiencing the same issue with restify(v7.3.0) + superagent(v3.8.3). `asyncHooks: false` works.

**Comment by JCMais** [comment]
Pretty sure it's the same issue @watson mentioned above, running the code at https://github.com/elastic/apm-agent-nodejs/issues/897#issuecomment-471520167 on Node.js 14.6 correctly creates the transaction.

![image](https://user-images.githubusercontent.com/1041647/91230572-27172300-e702-11ea-8891-c391694fbae7.png)
